### PR TITLE
Implemented skipping apicast service port reconcile based on annotation

### PR DESCRIPTION
**Pull request by bryankeane0**
This will resolve [THREESCALE-8385](https://issues.redhat.com/browse/THREESCALE-8385)

### Verification
- Deploy 3scale
- Add the annotation to the APIManager `marin3r.3scale.net/disable-apicast-service-port-reconcile: "true"`
- Confirm apicast service gateway port is 8443
- Made any change to apicast-production/staging Deployment Config to trigger reconciliation, like add a random annotation value `test-annotation: 'value'`
- Check apicast service that the gateway port stayed 8443 and hasn't changed to 8080

**Note**: _The above steps should be repeated with the annotation set to `'false'` or no annotation present at all and check that port changes upon reconciliation._

### CODE REVIEW

**Review by eguzki — CHANGES_REQUESTED** [review]
When the annotation is added, the operator does even create the apicast staging and production services. 

Use one mutator or another depending on the annotations. Something like:

```go
disableApicastPortReconcile := "false"
apiManagerAnnotations := r.apiManager.GetAnnotations()
if apiManagerAnnotations == nil {
	apiManagerAnnotations = make(map[string]string)
}
if val, ok := apiManagerAnnotations["apps.3scale.net/disable-apicast-service-reconciler"]; ok {
	disableApicastPortReconcile = val
}

apicastServiceMutator := reconcilers.CreateOnlyMutator
if disableApicastPortReconcile == "false" {
	apicastServiceMutator = reconcilers.ServicePortMutator
}

// Staging Service
err = r.ReconcileService(apicast.StagingService(), apicastServiceMutator)
if err != nil {
	return reconcile.Result{}, err
}

// Production Service
err = r.ReconcileService(apicast.ProductionService(), apicastServiceMutator)
if err != nil {
	return reconcile.Result{}, err
}

```

I would name the annotation
```
apps.3scale.net/disable-apicast-service-reconciler
```

This needs to be documented. I suggest to add some doc in https://github.com/3scale/3scale-operator/blob/master/doc/apimanager-reference.md#apimanager

The apicast reconciler has unittests and it should not be difficult to add one unit test that checks that with the annotation in place, running the reconciler does not update the apicast services.

File: pkg/3scale/amp/operator/apicast_reconciler.go [hunk 1]
```diff
@@ -109,16 +110,26 @@ func (r *ApicastReconciler) Reconcile() (reconcile.Result, error) {
 		return reconcile.Result{}, err
 	}
 
-	// Staging Service
-	err = r.ReconcileService(apicast.StagingService(), reconcilers.ServicePortMutator)
-	if err != nil {
-		return reconcile.Result{}, err
+	// Don't reconcile apicast production & staging service ports to allow RHOAM to handle service ports for RateLimiting
```
Comment (eguzki) [inline pkg/3scale/amp/operator/apicast_reconciler.go:113]: this comment should be generic, not RHOAM related

File: pkg/3scale/amp/operator/apicast_reconciler.go [hunk 2]
```diff
@@ -109,16 +110,26 @@ func (r *ApicastReconciler) Reconcile() (reconcile.Result, error) {
...
+
+	if metav1.HasAnnotation(r.apiManager.ObjectMeta, "marin3r.3scale.net/disable-apicast-service-port-reconcile") {
+		disableApicastPortReconcile = r.apiManager.Annotations["marin3r.3scale.net/disable-apicast-service-port-reconcile"]
 	}
 
-	// Production Service
-	err = r.ReconcileService(apicast.ProductionService(), reconcilers.ServicePortMutator)
-	if err != nil {
-		return reconcile.Result{}, err
+	// Reconcile service ports as normal if the annotation is false - otherwise skip this reconcile to let RHOAM handle apicast service port configuration
```
Comment (eguzki) [inline pkg/3scale/amp/operator/apicast_reconciler.go:120]: this comment should be generic, not RHOAM related

**Comment by bryankeane0** [comment]
@eguzki thank you for these comments, I'll implement these changes and add some test coverage and update the docs :)

File: pkg/3scale/amp/operator/apicast_reconciler.go [hunk 3]
```diff
@@ -109,14 +109,28 @@ func (r *ApicastReconciler) Reconcile() (reconcile.Result, error) {
...
+		apiManagerAnnotations = make(map[string]string)
+	}
+	if val, ok := apiManagerAnnotations["apps.3scale.net/disable-apicast-service-reconciler"]; ok {
+		disableApicastPortReconcile = val
+	}
+
+	apicastServiceMutator := reconcilers.CreateOnlyMutator
+	if disableApicastPortReconcile == "false" {
+		apicastServiceMutator = reconcilers.ServicePortMutator
+	}
```
Comment (KevFan) [inline pkg/3scale/amp/operator/apicast_reconciler.go:124]:
```suggestion
	apicastServiceMutator := r.getApiCastServiceMutator()
```

```
func (r ApicastReconciler) getApiCastServiceMutator() reconcilers.MutateFn {
	disableApicastPortReconcile := "false"
	apiManagerAnnotations := r.apiManager.GetAnnotations()
	if apiManagerAnnotations == nil {
		apiManagerAnnotations = make(map[string]string)
	}
	if val, ok := apiManagerAnnotations["apps.3scale.net/disable-apicast-service-reconciler"]; ok {
		disableApicastPortReconcile = val
	}

	apicastServiceMutator := reconcilers.CreateOnlyMutator
	if disableApicastPortReconcile == "false" {
		apicastServiceMutator = reconcilers.ServicePortMutator
	}

	return apicastServiceMutator
}
```

Minor suggestion, but would be nice to put this to it's own function and adding a unit test for testing the returned mutator :thinking:

Comment (bryankeane0) [reply pkg/3scale/amp/operator/apicast_reconciler.go:124]: Thanks Kevin, adding this now

**Comment by eguzki** [comment]
let me know when you are done and request re-review

File: doc/apimanager-reference.md [hunk 4]
```diff
@@ -69,22 +69,23 @@ Generated using [github-markdown-toc](https://github.com/ekalinin/github-markdow
...
+| DisableApicastServiceReconciler | `disableApicastServiceReconciler` | bool | No | `false` | When false, 3scale API management solution will reconcile apicast services as normal. Setting this to true will stop 3scale from reconcilin apicast service ports                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
```
Comment (eguzki) [inline doc/apimanager-reference.md:80]: This is part of the `spec` and you have defined an annotation which is part of `metadata`. It need it's own section that you need to  create as it is the first time we document an annotation for the APIManager

**Review by eguzki — CHANGES_REQUESTED** [review]
Looking very good.

Small changes requested and should be ready to be merged

File: doc/apimanager-reference.md [hunk 5]
```diff
@@ -7,55 +7,777 @@ One APIManager custom resource per project is allowed.
...
+  * [system-redis](#system-redis)
+  * [system-seed](#system-seed)
+  * [zync](#zync)
+  * [fileStorage-S3-credentials-secret](#filestorage-s3-credentials-secret)
+  * [system-smtp](#system-smtp)
+* [Default APIManager components compute resources](#default-apimanager-components-compute-resources)
+
+Generated using [github-markdown-toc](https://github.com/ekalinin/github-markdown-toc)
+
+## APIManager
```
Comment (eguzki) [inline doc/apimanager-reference.md:63]: What is this change? It seems to be duplicated

File: doc/apimanager-reference.md [hunk 6]
```diff
@@ -7,55 +7,777 @@ One APIManager custom resource per project is allowed.
...
+| SystemSpec  | `system`  | \*SystemSpec  | No | See [SystemSpec](#SystemSpec) reference | Spec of the System part |
+| ZyncSpec    | `zync`    | \*ZyncSpec    | No | See [ZyncSpec](#ZyncSpec) reference | Spec of the Zync part    |
+| HighAvailabilitySpec | `highAvailability` | \*HighAvailabilitySpec | No | **[DEPRECATED**] See [ExternalComponentsSpec](#ExternalComponentsSpec) reference | |
+| ExternalComponentsSpec | `externalComponents` | \*ExternalComponentsSpec | No | See [ExternalComponentsSpec](#ExternalComponentsSpec) reference | Spec of the ExternalComponentsSpec part |
+| PodDisruptionBudgetSpec | `podDisruptionBudget` | \*PodDisruptionBudgetSpec | No | See [PodDisruptionBudgetSpec](#PodDisruptionBudgetSpec) reference | Spec of the PodDisruptionBudgetSpec part |
+| MonitoringSpec | `monitoring` | \*MonitoringSpec | No | Disabled | [MonitoringSpec](#MonitoringSpec) reference |
+
+### APIManagerMetaData
+
+| **Annotations**  |  **json/yaml field**   |  **Type** | **Required** | **Default value** | **Description**      |
```
Comment (eguzki) [inline doc/apimanager-reference.md:92]: Annotations are always `map[string]string`. Proposed columns:
* Annotation
* Name
* Default value 
* Description

You documented that the default value is `true`. I would say it is `false`.

**Review by eguzki — CHANGES_REQUESTED** [review]

File: doc/apimanager-reference.md (hunk 5 above)
Comment (bryankeane0) [reply doc/apimanager-reference.md:62]: Hmm, possibly a plugin I wasn't aware I had on my IDE. removed it

Comment (eguzki) [reply doc/apimanager-reference.md:62]: I still see duplicated content

Comment (bryankeane0) [reply doc/apimanager-reference.md:64]: ah sorry, It's just me adding the metadata section to the toc

File: doc/apimanager-reference.md [hunk 7]
```diff
@@ -7,65 +7,67 @@ One APIManager custom resource per project is allowed.
...
+  * [system-smtp](#system-smtp)
 * [Default APIManager components compute resources](#default-apimanager-components-compute-resources)
 
 Generated using [github-markdown-toc](https://github.com/ekalinin/github-markdown-toc)
 
 ## APIManager
 
-| **Field** | **json/yaml field**| **Type** | **Required** | **Description** |
-| --- | --- | --- | --- | --- |
+| **Field** | **json/yaml field**| **Type** | **Required** | **Description**                                 |
```
Comment (eguzki) [inline doc/apimanager-reference.md:66]: Why this change? This was correct as it was

File: doc/apimanager-reference.md [hunk 8]
```diff
@@ -7,65 +7,67 @@ One APIManager custom resource per project is allowed.
...
+| --- | --- | --- | --- |-------------------------------------------------|
 | Spec | `spec` | [APIManagerSpec](#APIManagerSpec) | Yes | The specfication for APIManager custom resource |
-| Status | `status` | [APIManagerStatus](#APIManagerStatus) | No | The status for the custom resource |
+| Status | `status` | [APIManagerStatus](#APIManagerStatus) | No | The status for the custom resource              |
```
Comment (eguzki) [inline doc/apimanager-reference.md:69]: why this change in `status` row?

File: doc/apimanager-reference.md [hunk 9]
```diff
@@ -86,6 +88,14 @@ Generated using [github-markdown-toc](https://github.com/ekalinin/github-markdow
 | PodDisruptionBudgetSpec | `podDisruptionBudget` | \*PodDisruptionBudgetSpec | No | See [PodDisruptionBudgetSpec](#PodDisruptionBudgetSpec) reference | Spec of the PodDisruptionBudgetSpec part |
 | MonitoringSpec | `monitoring` | \*MonitoringSpec | No | Disabled | [MonitoringSpec](#MonitoringSpec) reference |
 
+### APIManagerMetaData
+
+| **Annotations**  | **Name** | **Default value** | **Description**                                                                  |
```
Comment (eguzki) [inline doc/apimanager-reference.md:93]: it should be:

```
| **Annotations** | **Name** | **Default value** | **Description** |
| --- | --- | --- | --- |
```

**Review by eguzki — CHANGES_REQUESTED** [review]
almost done. Little comments dropped

File: doc/apimanager-reference.md (hunk 9 above)
Comment (bryankeane0) [reply doc/apimanager-reference.md]: Apologies, these came from a plugin I didn't know I had on my IDE that reformats code, I've fixed all of the above. Thanks for checking it again :)

**Review by eguzki — APPROVED** [review]
looking good

when the tests pass, it will be merged
